### PR TITLE
Flash Firmware on the Circuit Playground Classic

### PR DESCRIPTION
**Pull request by epeach**
Relies on: https://github.com/code-dot-org/browser/pull/52
Currently behind flag '?enableExperiments=flash-classic'
When the flag is on and the user goes from to 'studio.code.org/maker/setup', there is an extra setup step that lets the user know they should update the board firmware, with a button to trigger the firmware update. This step does not appear for any other board.
With flag and classic board plugged in:
Screenshot from 2019-08-09 14-20-51

After clicking button, while flashing:
Screenshot from 2019-08-09 14-22-29

Successful firmware flash:
Screenshot from 2019-08-09 14-21-06

Next steps:
* Only display this validation step when we detect the firmware is out of date.

Open questions:
*When the user flashes the board, how should the UI respond? Should the page refresh or replay the 'celebrate actions' or other?

### CODE REVIEW

File: apps/src/lib/kits/maker/ui/SetupChecklist.jsx [hunk 1]
```diff
@@ -314,6 +338,25 @@ export default class SetupChecklist extends Component {
...
             .{this.contactSupport()}
           </ValidationStep>
+          {experiments.isEnabled('flash-classic') &&
+            this.state.boardTypeDetected === BOARD_TYPE.CLASSIC && (
+              <ValidationStep
+                stepStatus={this.state[STATUS_BOARD_FIRMWARE]}
+                stepName="Board firmware up-to-date"
+              >
+                <div>
+                  This board is not using the most up-to-date version of the
```
Comment (epeach) [inline apps/src/lib/kits/maker/ui/SetupChecklist.jsx]: Temp text until we get final text from Ryan's design

Comment (joshlory) [reply apps/src/lib/kits/maker/ui/SetupChecklist.jsx]: Localization? (once we get the final text..)

**Comment by joshlory** [comment]
giphy

File: apps/src/lib/kits/maker/ui/SetupChecklist.jsx (hunk 1 above)
Comment (epeach) [reply apps/src/lib/kits/maker/ui/SetupChecklist.jsx]: I should probably localize now, so I don't forget to do it later :laughing:

**Comment by epeach** [comment]
@joshlory Based on our discussion about keeping this behind a flag while moving forward with the detection code, PTAL!

File: apps/src/lib/kits/maker/ui/SetupChecklist.jsx [hunk 2]
```diff
@@ -140,6 +147,20 @@ export default class SetupChecklist extends Component {
...
+   * Update the firmware on the attached board. Currently, only the CPClassic can be flashed.
+   * @return {Promise}
+   */
+  updateBoardFirmware() {
+    this.setState({[STATUS_BOARD_FIRMWARE]: Status.ATTEMPTING});
+    return window.MakerBridge.flashBoardFirmware({
+      boardName: 'circuit-playground-classic',
+      hexPath:
+        'https://s3.amazonaws.com/downloads.code.org/maker/CircuitPlaygroundFirmata.ino.circuitplay32u4.hex',
+      checksum: '766c4fad9088037ab4839b18292be8b1'
```
Comment (joshlory) [inline apps/src/lib/kits/maker/ui/SetupChecklist.jsx]: Not blocking merge of this experiment, but don't we want to point to a latest.json file that includes the checksum and a link to the hex file?

File: apps/src/lib/kits/maker/ui/SetupChecklist.jsx [hunk 3]
```diff
@@ -314,6 +335,21 @@ export default class SetupChecklist extends Component {
             </a>
             .{this.contactSupport()}
           </ValidationStep>
+          {experiments.isEnabled('flash-classic') &&
+            this.state.boardTypeDetected === BOARD_TYPE.CLASSIC && (
+              <ValidationStep
+                stepStatus={this.state[STATUS_BOARD_FIRMWARE]}
+                stepName="Board firmware up-to-date"
```
Comment (joshlory) [inline apps/src/lib/kits/maker/ui/SetupChecklist.jsx]: Nit: i18n

File: apps/src/lib/kits/maker/ui/SetupChecklist.jsx [hunk 4]
```diff
@@ -15,21 +15,26 @@ import {
   isLinux
 } from '../util/browserChecks';
 import ValidationStep, {Status} from '../../../ui/ValidationStep';
+import {BOARD_TYPE} from '../CircuitPlaygroundBoard';
+import experiments from '../../../../util/experiments';
```
Comment (joshlory) [inline apps/src/lib/kits/maker/ui/SetupChecklist.jsx]: Tiny style nit, prefer the `@cdo/apps` alias for deeply-nested dependencies:
```suggestion
import experiments from '@cdo/apps/util/experiments';
```

**Review by joshlory — APPROVED** [review]
A few nits, but nothing blocking merge of this experiment!

File: apps/src/lib/kits/maker/ui/SetupChecklist.jsx (hunk 2 above)
Comment (epeach) [reply apps/src/lib/kits/maker/ui/SetupChecklist.jsx]: Updated

File: apps/src/lib/kits/maker/ui/SetupChecklist.jsx [hunk 5]
```diff
@@ -17,6 +17,8 @@ import {
 import ValidationStep, {Status} from '../../../ui/ValidationStep';
 import {BOARD_TYPE} from '../CircuitPlaygroundBoard';
 import experiments from '@cdo/apps/util/experiments';
+import _ from 'lodash';
+import yaml from 'js-yaml';
```
Comment (joshlory) [inline apps/src/lib/kits/maker/ui/SetupChecklist.jsx:21]: Tiny nit: could make the manifest a JSON file so we don't need to pull in a library to parse it.

**Pull request merged**